### PR TITLE
azure-pipelines: test CI setup

### PR DESCRIPTION
**Pull request by iMichka**

### CODE REVIEW

**Comment by iMichka** [comment]
@ladislas @sjackman I am really struggling to get this working. I added a lot of debug output to try to understand what is happening with the start/end sha1.

I am using the `--no-pull` option too, not sure I should? It is totally unclear to me in what initial state the repository should be?:
- should the branch be checked out?
- should the pr be checked out?
- should master be updated to origin/master? Should the pr be rebased on top of it?
- or should nothing be done, and brew test-bot will take care of the setup for me?

```
Testing tap homebrew/core with:
  origin/master   3cbba5237c (Merge branch homebrew/master into linuxbrew/master)
  HEAD            197040e951 (azure-pipelines: test CI setup)
  diff_start_sha1 197040e951bf380ea0da556839ebe12730b45b2c
  diff_end_sha1   197040e951

Formula changes to be tested:
  added formulae    (empty)
  modified formulae (empty)
  deleted formulae  (empty)
```

I am also wondering why we are mixing short and long sha1?

**Comment by ladislas** [comment]
@iMichka I use the following before everything else to setup the environment. @sjackman suggested it here: https://github.com/Homebrew/homebrew-test-bot/pull/256#issuecomment-491990986

```
      set -ex
      if [[ -n "$SYSTEM_PULLREQUEST_PULLREQUESTNUMBER" ]]; then
        git fetch origin "master:master" "pull/$SYSTEM_PULLREQUEST_PULLREQUESTNUMBER/head:pr"
        git checkout pr
      fi
```

https://github.com/ladislas/homebrew-hello-world/blob/master/.azure-ci/setup-ci-env.yml#L9-L13

> I am also wondering why we are mixing short and long sha1?

Good question, I don't know but I'll look into it :)

**Comment by ladislas** [comment]
plus in your case it seems to be the `brew doctor` that's failing

```console
==> brew doctor
==> FAILED
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Homebrew/linuxbrew-core is not on the master branch.

Check out the master branch by running:
  git -C "$(brew --repo homebrew/core)" checkout master
```

https://dev.azure.com/Homebrew/Homebrew/_build/results?buildId=9901&view=logs&jobId=12f1170f-54f2-53f3-20dd-22fc7dff55f9&taskId=356bb04c-cb4a-5f04-82ca-d3b102917eba&lineStart=124&lineEnd=133&colStart=1&colEnd=56

**Comment by ladislas** [comment]
> I am also wondering why we are mixing short and long sha1?

it's because the final value of `diff_start_sha1` is calculated with a `git merge-base` that does not create a `--short` version of the hash.

**Comment by iMichka** [comment]
I tried:

```
set -ex
if [[ -n "$SYSTEM_PULLREQUEST_PULLREQUESTNUMBER" ]]; then
  git fetch origin "master:master" "pull/$SYSTEM_PULLREQUEST_PULLREQUESTNUMBER/head:pr"
  git checkout pr
fi
```

and I got:

> fatal: Refusing to fetch into current branch refs/heads/master of non-bare repository
error: pathspec 'pr' did not match any file(s) known to git

**Comment by ladislas** [comment]
Can you show me?
It seems that you are running this from `master` hence the error.

**Comment by sjackman** [comment]
> I am using the --no-pull option too, not sure I should?

Yes.

> It is totally unclear to me in what initial state the repository should be?:
> 
> should the branch be checked out?

A branch should be checked out.

> should the pr be checked out?

Yes, either `pull/1234/head` should be checked out.

> should master be updated to origin/master?

Yes.

> Should the pr be rebased on top of it?

For now no. Eventually `pull/1234/merge` will be checked out, but that's not possible now due to an issue with `brew test-bot` triggering the special merge logic.

> or should nothing be done, and brew test-bot will take care of the setup for me?

That's not possible yet, but we may get there. Mike prefers this solution.

**Comment by sjackman** [comment]
> fatal: Refusing to fetch into current branch refs/heads/master of non-bare repository
error: pathspec 'pr' did not match any file(s) known to git

That's because `master` is checked out. AzureCI by default has you on a detached head (which I think is a bit weird) of `pull/1234/merge`.

**Comment by sjackman** [comment]
> I am really struggling to get this working. I added a lot of debug output to try to understand what is happening with the start/end sha1.

I suggest merging PR #13265 so that you have a working starting point. It had only one easily-resolvable `brew tests` failure and is based on the known-good configuration of Brewsci/bio.

**Comment by iMichka** [comment]
> I suggest merging PR #13265 so that you have a working starting point

Yes but the real implementation will be in docker, so it does not really help me. I will do that in case I see no way forward here.

> That's because master is checked out. AzureCI by default has you on a detached head (which I think is a bit weird) of pull/1234/merge.

Agreed, I have seen that. So I will just try to checkout a new empty branch to be able to run `git fetch origin "master:master" "pull/$SYSTEM_PULLREQUEST_PULLREQUESTNUMBER/head:pr"`

**Comment by sjackman** [comment]
>> I suggest merging PR #13265 so that you have a working starting point

> Yes but the real implementation will be in docker, so it does not really help me. I will do that in case I see no way forward here.

In my opinion it's a step in the right direction. Having a known good starting point is better than no starting point. It's also nice to be able to see the diff from a without-Docker implementation to a with-Docker implementation. Note that `brew test-bot` started without-Docker and moved to with-Docker. You could see the diff over there.

**Comment by iMichka** [comment]
I feel that I am almost there. Also, the example in #13265 does not build bottles, it runs the tests on brew. Not sure why though.

I checked out a "test" branch so the I could run:
```
git fetch origin "master:master" "pull/$SYSTEM_PULLREQUEST_PULLREQUESTNUMBER/head:pr"
git checkout pr
```

This worked. Now I can print some output (basically the same code executed by the test bot):

```
current_sha1=$(git rev-parse --short HEAD)
diff_end_sha1=$current_sha1
diff_start_sha1=$(git rev-parse --short "$SYSTEM_PULLREQUEST_TARGETBRANCH")
diff_start_sha1_new=$(git merge-base $diff_start_sha1 $diff_end_sha1)
echo $current_sha1
echo $diff_start_sha1
echo $diff_end_sha1
echo $diff_start_sha1_new

f7a46c1bc3
3cbba5237c
f7a46c1bc3
95ebc5335fd26d97fdf85f5f0b8bf5a2fe037dda
```

`f7a46c1bc3` is the commit  of the PR. `diff_end_sha1` is therefor correct. `3cbba5237c` is the latest commit on origin/master. `95ebc5335fd26d97fdf85f5f0b8bf5a2fe037dda` is the parent commit from which the branch was created, somewhere on master. `diff_start_sha1` should be set to that value when we call `git merge-base $diff_start_sha1 $diff_end_sha1`.

Now, in the final output, `diff_start_sha1` is still set as the commit of the PR. Normally that value should be `95ebc5335fd26d97fdf85f5f0b8bf5a2fe037dda`.

Testing tap homebrew/core with:
  origin/master   3cbba5237c (Merge branch homebrew/master into linuxbrew/master)
  HEAD            f7a46c1bc3 (azure-pipelines: test CI setup)
  diff_start_sha1 f7a46c1bc36e7d981c87750fbbf77a0d48a116af
  diff_end_sha1   f7a46c1bc3

File: azure-pipelines.yml [hunk 1]
```diff
@@ -1,19 +1,61 @@
...
+    git config --global user.name LinuxbrewTestBot
+    git config --global user.email testbot@linuxbrew.sh
+    cd /home/linuxbrew/.linuxbrew/Homebrew
+    sudo git fetch origin --tags
+    sudo git reset --hard origin/master
+  displayName: Setup Homebrew
+- bash: |
+    sudo rm -rf /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core
+    sudo cp -a $(Build.Repository.LocalPath) /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core
+  displayName: Copy homebre-core
```
Comment (sjackman) [inline azure-pipelines.yml]:
```suggestion
  displayName: Copy homebrew-core
```

File: azure-pipelines.yml [hunk 2]
```diff
@@ -1,19 +1,61 @@
...
+    current_sha1=$(git rev-parse --short HEAD)
+    diff_end_sha1=$current_sha1
+    diff_start_sha1=$(git rev-parse --short "$SYSTEM_PULLREQUEST_TARGETBRANCH")
+    diff_start_sha1_new=$(git merge-base $diff_start_sha1 $diff_end_sha1)
+    echo $current_sha1
+    echo $diff_start_sha1
+    echo $diff_end_sha1
+    echo $diff_start_sha1_new
+    echo "4 -----------"
+    sudo /home/linuxbrew/.linuxbrew/bin/brew test-bot --no-pull --tap=homebrew/core --bintray-org=linuxbrew --git-name=LinuxbrewTestBot --git-email=testbot@linuxbrew.sh --keep-old
```
Comment (sjackman) [inline azure-pipelines.yml]:
```suggestion
    /home/linuxbrew/.linuxbrew/bin/brew test-bot --keep-old --tap=homebrew/core --bintray-org=linuxbrew --git-name=LinuxbrewTestBot --git-email=testbot@linuxbrew.sh
```
`sudo` is clearing all of your environment variables.

Comment (sjackman) [reply azure-pipelines.yml]: `--no-pull` is the default when `brew test-bot` sees that it's running in CI.

Comment (iMichka) [reply azure-pipelines.yml]: Regarding `--no-pull`, I did not realise it was set on CI. I removed it.

I also removed the `sudo` call. Let's see what happens.

Comment (sjackman) [reply azure-pipelines.yml]: Here's where `--no-pull` is set if you're curious.
https://github.com/Homebrew/homebrew-test-bot/blob/ddd1b9cbefc66422da77073f5fb1288f1ccd9728/cmd/brew-test-bot.rb#L1618

**Comment by iMichka** [comment]
This is why I was calling brew test-bot with sudo:

```
Homebrew/homebrew-test-bot ddd1b9c (Merge pull request #262 from ladislas/feature/puts-added-modified-formulae)
ARGV: --tap=homebrew/core --bintray-org=linuxbrew --git-name=LinuxbrewTestBot --git-email=testbot@linuxbrew.sh --keep-old --verbose --ci-auto --no-pull --ci-pr --cleanup --test-default-formula --junit

==> sudo rm -rf /usr/local/include/node

==> brew untap linuxbrew/extra
Error: Permission denied @ rb_sysopen - /home/vsts_azpcontainer/.cache/Homebrew/.descriptions.json20190522-2862-vaczf8
Untapping linuxbrew/extra...
==> FAILED
Error: Permission denied @ dir_s_mkdir - /home/linuxbrew/.linuxbrew/lib/pkgconfig
```

**Comment by sjackman** [comment]
```
Error: 1 problem in 1 formula detected
hello:
  * Incorrect file permissions (777): chmod 644 /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/hello.rb
==> FAILED
```
https://dev.azure.com/Homebrew/Homebrew/_build/results?buildId=9938&view=logs&jobId=12f1170f-54f2-53f3-20dd-22fc7dff55f9&taskId=356bb04c-cb4a-5f04-82ca-d3b102917eba&lineStart=481&lineEnd=482&colStart=1&colEnd=1

**Comment by iMichka** [comment]
@sjackman here we go. The issue was with `sudo`, and `sudo -E` fixed it. Never had thought that this may be an issue. I merged the first version as a test, it builds bottles. Let's see how this behaves in the next days. I'll probably make a test to build `llvm` or `rust`, just to see if we do not hit any issues.

We will need to think about how to upload the bottles now :)

**Comment by ladislas** [comment]
> We will need to think about how to upload the bottles now :)

If you don't need to build for multiple versions of Linux, you can can do it in a following step.

If not, you'll need to save the artifacts before downloading them in another job.

**Pull request merged**